File: leaction-platform/examples/cms-consumer-noticias.py
"""
Exemplo Flask (PanelDX) — espelho do consumidor inove4us.

Copie para LeAction_SysF e registre:

    from cms_noticias_routes import cms_noticias_bp  # ou cole a rota abaixo
    app.register_blueprint(cms_noticias_bp)

Env:
  ACTION_HUB_API_URL=http://127.0.0.1:4001
  CMS_CACHE_TTL_SEC=480
"""

from __future__ import annotations

import logging
import os
import threading
import time

import requests
from flask import Blueprint, jsonify, request

logger = logging.getLogger(__name__)

CACHE_TTL_SEC = int(os.environ.get("CMS_CACHE_TTL_SEC", "480"))
HUB_TIMEOUT_SEC = float(os.environ.get("CMS_HUB_TIMEOUT_SEC", "3.5"))

_lock = threading.Lock()
_cache: dict = {}

cms_noticias_bp = Blueprint("cms_noticias", __name__)


def _hub_base() -> str:
    return (
        os.environ.get("ACTION_HUB_API_URL")
        or os.environ.get("HUB_API_URL")
        or "http://127.0.0.1:4001"
    ).strip().rstrip("/")


def fetch_posts(sistema: str = "paneldx", limit: int = 5) -> list:
    key = f"{sistema}:{limit}"
    now = time.time()
    with _lock:
        entry = _cache.get(key)
        if entry and (now - entry["fetched_at"]) < CACHE_TTL_SEC:
            return list(entry["posts"])

    url = f"{_hub_base()}/api/cms/posts"
    try:
        res = requests.get(
            url,
            params={"sistema_destino": sistema, "limit": str(limit)},
            timeout=HUB_TIMEOUT_SEC,
        )
        res.raise_for_status()
        data = res.json() if res.content else {}
        posts = data.get("posts") if isinstance(data, dict) else None
        if not isinstance(posts, list):
            posts = []
        with _lock:
            _cache[key] = {"posts": posts, "fetched_at": now}
        return list(posts)
    except Exception as exc:
        logger.warning("[cms] Hub indisponível: %s", exc)
        with _lock:
            stale = _cache.get(key)
            if stale:
                return list(stale["posts"])
        return []


@cms_noticias_bp.get("/api/noticias")
def list_noticias():
    sistema = (request.args.get("sistema_destino") or "paneldx").strip() or "paneldx"
    try:
        limit = int(request.args.get("limit") or "5")
    except (TypeError, ValueError):
        limit = 5
    limit = max(1, min(limit, 50))
    return jsonify(fetch_posts(sistema, limit)), 200

File: leaction-platform/examples/cms-consumer-noticias.js
/**
 * Exemplo Node (PanelDX / BFF) — GET /api/noticias ou /bff/noticias
 *
 * Em produção no PanelDX, /api/* costuma ir para o Flask no ALB.
 * Prefira registrar no Express como /bff/noticias, ou espelhar a mesma
 * lógica no Flask (ver inove4us/backend/cms_noticias_routes.py).
 *
 * Uso:
 *   const { createNoticiasHandler } = require('./cms-consumer-noticias');
 *   app.get('/bff/noticias', createNoticiasHandler({ sistema: 'paneldx' }));
 */

'use strict';

const CACHE_TTL_MS = Number(process.env.CMS_CACHE_TTL_MS || 8 * 60 * 1000);
const HUB_TIMEOUT_MS = Number(process.env.CMS_HUB_TIMEOUT_MS || 3500);

const cache = new Map(); // key -> { posts, fetchedAt }

function hubBase() {
  return (
    process.env.ACTION_HUB_API_URL ||
    process.env.HUB_API_URL ||
    'http://127.0.0.1:4001'
  ).replace(/\/$/, '');
}

async function fetchFromHub(sistema, limit) {
  const url = new URL(`${hubBase()}/api/cms/posts`);
  url.searchParams.set('sistema_destino', sistema);
  url.searchParams.set('limit', String(limit));

  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), HUB_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      signal: ctrl.signal,
      headers: { Accept: 'application/json' },
    });
    if (!res.ok) throw new Error(`Hub HTTP ${res.status}`);
    const data = await res.json();
    return Array.isArray(data?.posts) ? data.posts : [];
  } finally {
    clearTimeout(timer);
  }
}

function createNoticiasHandler(options = {}) {
  const defaultSistema = options.sistema || 'paneldx';
  const defaultLimit = options.limit || 5;

  return async function noticiasHandler(req, res) {
    const sistema = String(req.query.sistema_destino || defaultSistema).trim() || defaultSistema;
    let limit = parseInt(String(req.query.limit || defaultLimit), 10);
    if (!Number.isFinite(limit) || limit < 1) limit = defaultLimit;
    if (limit > 50) limit = 50;

    const key = `${sistema}:${limit}`;
    const now = Date.now();
    const hit = cache.get(key);
    if (hit && now - hit.fetchedAt < CACHE_TTL_MS) {
      return res.status(200).json(hit.posts);
    }

    try {
      const posts = await fetchFromHub(sistema, limit);
      cache.set(key, { posts, fetchedAt: now });
      return res.status(200).json(posts);
    } catch (err) {
      console.warn('[cms] Hub indisponível:', err?.message || err);
      if (hit?.posts) return res.status(200).json(hit.posts);
      return res.status(200).json([]);
    }
  };
}

module.exports = { createNoticiasHandler, fetchFromHub };
